device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/ProxiedSSLSocket.java

```diff
@@ -0,0 +1,204 @@
+
+        String connectResponse = in.readHttpConnectResponse();
+
+        String[] connectResponseLines = connectResponse.split("\r\n");
+
+        int connectResponseStart = 0;
+        while (connectResponseLines[connectResponseStart].isEmpty())
+        {
+            connectResponseStart++;
+        }
+
+        //Expects the same http version in the response as the request
+        String firstLine = connectResponseLines[connectResponseStart];
+        if (!firstLine.startsWith(HTTP_VERSION_1_1))
+        {
+            tunnel.close();
+            throw new IOException(String.format("Unable to tunnel through %s:%d.  Expected first response line to start with %s, but proxy returns \"%s\"", host, port, HTTP_VERSION_1_1, firstLine));
+        }
+
+        String[] replyStrParts = firstLine.split(" ");
+        if (replyStrParts.length < 2)
+        {
+            tunnel.close();
+            throw new IOException(String.format("Unable to tunnel through %s:%d. Expected proxy response to CONNECT to contain a space between http version and status code, but was %s", host, port, firstLine));
+        }
+
+        int connectResponseStatusCode;
+        try
+        {
+            connectResponseStatusCode = Integer.parseInt(replyStrParts[1]);
+        }
+        catch (NumberFormatException e)
+        {
+            tunnel.close();
+            throw new IOException(String.format("Unable to tunnel through %s:%d. Expected proxy response to CONNECT to contain a status code but status code could not be parsed. Response was %s", host, port, firstLine));
+        }
+
+        if (connectResponseStatusCode <= 199 || connectResponseStatusCode >= 300)
+        {
+            tunnel.close();
+            throw new IOException(String.format("Unable to tunnel through %s:%d. Expected proxy response to CONNECT to return status code 2XX but status code was %d", host, port, connectResponseStatusCode));
+        }
+
+        log.trace("HTTP proxy responded to connect request with status {}, so the proxy connect was successful", connectResponseStatusCode);
+    }
+
+    private interface ProxiedSSLSocketNonDelegatedFunctions
+    {
+        void connect(SocketAddress socketAddress, int timeout) throws IOException;
+        void connect(SocketAddress socketAddress) throws IOException;
+        void close() throws IOException;
+    }
+
+    @RequiredArgsConstructor
+    class HttpConnectResponseReader
+    {
+        private boolean alreadyRead = false;
+        @NonNull private InputStream inputStream;
+        @NonNull private Charset byteEncoding;
+
+        String readHttpConnectResponse() throws IOException
+        {
+            if (alreadyRead)
+            {
+                throw new IOException("Http connect response has already been read");
+            }
+
+            ByteArrayOutputStream httpLineOutputStream = new ByteArrayOutputStream();
+            char[] mostRecentFourCharacters = new char[4];
```

Comment:
`LinkedList` will be the best fit if you want to keep 4 chars.

Reply:
Okay, I've switched to using a linked list instead